synapse/handlers/profile.py

```diff
@@ -168,6 +173,13 @@ async def set_displayname(
         if new_displayname == "":
             new_displayname = None
 
+        # Since the requester can get overwritten below, check if the real
+        # requester is shadow-banned.
+        if requester.shadow_banned:
+            # We randomly sleep a bit just to annoy the requester a bit.
+            await self.clock.sleep(random.randint(1, 10))
+            raise ShadowBanError()
+
```

Comment:
I'm not sure if this is necessary. Shouldn't we let shadow-banned users change their displaynames if they want? It won't make it into any rooms thanks to the check in `update_membership`.

Reply:
Hm....I was under the impression it would still end up in the room, but I think you're right. I'll need to test this.

Reply:
So the changes to `update_membership` won't catch this because those only apply to `invite` requests.